### ISSUE TITLE
Adding new presets for pytorch c++ frontend?

### ISSUE THREAD

**Issue by wumo** [bytedeco/javacpp-presets#623]
Recently `pytorch` release [preview version 1.0](https://github.com/pytorch/pytorch/releases/tag/v1.0rc1), which features the experimental `C++ frontend`. The pytorch C++ frontend is easier and more complete than `tensorflow`.
>The C++ frontend is a pure C++ interface to the PyTorch backend that follows the API and architecture of the established Python frontend. It is intended to enable research in high performance, low latency and bare metal C++ applications. It provides equivalents to torch.nn, torch.optim, torch.data and other components of the Python frontend. Here is a minimal side-by-side comparison of the two language frontends:

Python:
```python
import torch

model = torch.nn.Linear(5, 1)
optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
prediction = model.forward(torch.randn(3, 5))
loss = torch.nn.functional.mse_loss(prediction, torch.ones(3, 1))
loss.backward()
optimizer.step()
```
C++:
```c++
#include <torch/torch.h>

torch::nn::Linear model(5, 1);
torch::optim::SGD optimizer(model->parameters(), /*lr=*/0.1);
torch::Tensor prediction = model->forward(torch::randn({3, 5}));
auto loss = torch::mse_loss(prediction, torch::ones({3, 1}));
loss.backward();
optimizer.step();
```

Will you consider adding the pytorch c++ frontend presets?

**Comment by saudet** [comment]
AFAIK, that's what was known as Caffe2, and sure it make sense to want to
use it from Java.

**Comment by saudet** [comment]
BTW, @Neiko2002, is this something you would be also interested in having?

**Comment by Neiko2002** [comment]
@saudet To be honest I have never tried pytorch yet. Their python API looks similar to Java. A javacpp version of it might be easier to use than the official C++ API. Right now we deploy javacpp-tensorflow in production systems but train networks in python. Pytorch on the other hand has a simple C++ API to train and run a network. 

Looking at the [C++ API](https://pytorch.org/cppdocs/api/library_root.html#class-hierarchy) it does not feel it offers more than the Tensorflow API.

**Comment by saudet** [comment]
I think there is great potential in being able to (re)train on a platform that's more efficient than CPython, so I do hope this panes out. In any case, TensorFlow has obviously given up on that, let's see where PyTorch is willing to go! They used to promote Lua for this exact reason, although not Java for whatever political reason, so anyway IMO there is a chance they will redo something similar for C++11:
http://bytedeco.org/news/2015/03/14/java-meets-caffe/

@karllessard If you have any opinions about this, I would love to hear them!

**Comment by karllessard** [comment]
> @karllessard If you have any opinions about this, I would love to hear them!

For the moment no I don't but this is very interesting, thanks for the tip @saudet !

**Comment by saudet** [comment]
I've been looking at the the C++ API over the past couple of days, and it looks like a mess. It mixes types from Caffe2, ATen, and C10 with no clear understanding of how they fit together: https://github.com/pytorch/pytorch/issues/14850

Moreover, it's currently unable to train for MNIST using simple examples on either CPU or GPU:
https://discuss.pytorch.org/t/in-c-api-1-0-0-mnist-results-are-unstable/32823
https://github.com/pytorch/pytorch/issues/15522
I've personally confirmed those results by building from source both PyTorch 1.0.0 and current master, both with and without CUDA/cuDNN.

We should probably wait until all this gets cleared up and fixed before starting to map anything to Java...

**Comment by iwroldtan** [comment]
> Recently `pytorch` release [preview version 1.0](https://github.com/pytorch/pytorch/releases/tag/v1.0rc1), which features the experimental `C++ frontend`. The pytorch C++ frontend is easier and more complete than `tensorflow`.
> 
> > The C++ frontend is a pure C++ interface to the PyTorch backend that follows the API and architecture of the established Python frontend. It is intended to enable research in high performance, low latency and bare metal C++ applications. It provides equivalents to torch.nn, torch.optim, torch.data and other components of the Python frontend. Here is a minimal side-by-side comparison of the two language frontends:
> 
> Python:
> 
> ```python
> import torch
> 
> model = torch.nn.Linear(5, 1)
> optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
> prediction = model.forward(torch.randn(3, 5))
> loss = torch.nn.functional.mse_loss(prediction, torch.ones(3, 1))
> loss.backward()
> optimizer.step()
> ```
> 
> C++:
> 
> ```c++
> #include <torch/torch.h>
> 
> torch::nn::Linear model(5, 1);
> torch::optim::SGD optimizer(model->parameters(), /*lr=*/0.1);
> torch::Tensor prediction = model->forward(torch::randn({3, 5}));
> auto loss = torch::mse_loss(prediction, torch::ones({3, 1}));
> loss.backward();
> optimizer.step();
> ```
> 
> Will you consider adding the pytorch c++ frontend presets?

Could you give me a hand? I can not find the file, torch/torch.h.

**Comment by saudet** [comment]
@iwroldtan see https://pytorch.org/cppdocs/installing.html

**Comment by jxtps** [comment]
It looks like that specific instability issue has been resolved as https://github.com/pytorch/pytorch/issues/15522 is now closed. 

Looking at their release log they seem to have been pretty busy: https://github.com/pytorch/pytorch/releases

Though the C++ API is still labeled as being in beta: https://pytorch.org/cppdocs/

Might it make sense to package something up?

**Comment by saudet** [comment]
@jxtps It still looks like a mess to me. There's still very little information about what C10 is supposed to be, apart from this: https://github.com/pytorch/pytorch/wiki/Software-Architecture-for-c10 But if you're willing to understand what we need to map and how, that would be a good first step I think.

**Comment by jxtps** [comment]
Yeah, it seems like they've got a lot of history to deal with in that project. 

Judging by https://pytorch.org/cppdocs/ it looks like the C++ frontend (= `torch/torch.h`) and its dependencies (which I presume include ATen and Autograd but maybe not TorchScript or the C++ extensions?) could make sense. 

I would hope that wherever they end up with their c10 Tensor-library, it will be backwards compatible in the sense that 1. all relevant operations are preserved and 2. no operation subtly changes behavior but retains the same overall signature. So at some upgrade point there will be a massive number of compile errors as `at::Tensor` changes to `c10::Tensor` (or whatever), and the names of some methods change, but it's more of a "replace all", not "for every single call site figure out precisely what maps to what" type situation. 

I recognize that the correct approach here may very well be "wait another year".

**Comment by doofin** [comment]
Hi,as of 2021 the cpp frontend  seems stable enough,I think we can just look files at `libtorch/include/torch/csrc/api/include/torch/` and skip things like c10,Aten,etc.desperately want to have this bindings!

There are projects like `https://github.com/nazarblch/torch-scala`,but I haven't really understand this project.It uses cmake to generate sth which is confusing

**Comment by saudet** [comment]
There's been some work on this recently. @wmeddie Would you be able to make public what you have?

**Comment by wmeddie** [comment]
My fork is available https://github.com/bytedeco/javacpp-presets/compare/master...wmeddie:master  But need to give it a little bit of a clean up.  If other people are interested in helping get it over the finish line I can create a PR over the weekend I think.

**Comment by doofin** [comment]
@wmeddie Thanks for your nice job! I would be glad to help since I want to do this 2 years ago.. I have forked from your branch,will start a initial test build.

**Comment by saudet** [comment]
Hey everyone, I've finally taken some time to start working on this myself, and although there is still much to be done, the basics are there to get something like a simple training example for MNIST working on Linux, Mac, and Windows, with MKL and everything:
 * https://github.com/bytedeco/javacpp-presets/tree/actions/pytorch#sample-usage

Now, some important parts of the C++ API are probably missing and/or not functional, yet, and since it has a pretty large surface it would be a good idea to prioritize needs. Please let me know what you think we should work on next:
 * Enable more build options? Which ones?
 * Add integration with the Python API as well?
 * Map/fix more of the C++ API? Which classes, functions, variables, etc?
 * Port more examples from C++ to Java?
 * Do something else entirely?

In any case, please give it a try with the snapshots (instructions at http://bytedeco.org/builds/) and let me know what you think!
Thank you for your interest

/cc @HGuillemet @stu1130

**Comment by HGuillemet** [comment]
I gave a first try. It seems quite promising.
I tried to port some CNN model from Python and I faced the following problem: many standard modules use an `Options` class that makes use of the `TORCH_ARG` macro. This macro takes a type, a name and a value and defines a private instance variable initialized with the value, a setter returning `this` and a getter. 
This macro should be interpreted by the parser and generate a matching "builder-type" interface. Currently we cannot set these options.

Also it would be interesting to have some Java-ish replacement for the ExpandingArray template type. Can we use generics here ?

**Comment by saudet** [comment]
> I gave a first try. It seems quite promising.

Thanks for trying it out!

> I tried to port some CNN model from Python and I faced the following problem: many standard modules use an `Options` class that makes use of the `TORCH_ARG` macro. This macro takes a type, a name and a value and defines a private instance variable initialized with the value, a setter returning `this` and a getter.

`TORCH_ARG()` only seems to be used to define new options, which is not something users typically need to do. Could you show an example in C++ or in Python of what you're trying to do?

> This macro should be interpreted by the parser and generate a matching "builder-type" interface. Currently we cannot set these options.
> 
> Also it would be interesting to have some Java-ish replacement for the ExpandingArray template type. Can we use generics here ?

We can add/generate as many "helper" methods and classes as we want, sure, but there's no way to automate this kind of translation process from C++ to Java in a generic fashion, so that kind of thing is probably not going to be part of JavaCPP itself or of the presets for PyTorch, but of some higher-level project, such as JavaCV in the case of OpenCV, FFmpeg, etc, or TF Java in the case of TensorFlow.

**Comment by HGuillemet** [comment]
> Could you show an example in C++ or in Python of what you're trying to do?

The configuration of a Convolution module should look like this:
```
conv = new Conv2dImpl(
         new Conv2dOptions(numChannelsIn, numChannelsOut, new int[] { 3 })
             .stride(3)
             .padding(1)
             .bias(false)
           );
```
But the stride, padding, bias, etc... setters are not defined yet because declared in `torch/nn/options/conv.h` with the `TORCH_ARG` macro. Also the `new int[] { 3 }` is the kernel size and the Java presets only take a `LongPointer` as a replacement for a `ExpandingArray<2>`

**Comment by saudet** [comment]
I see, I'm not sure how we could make this work well, it uses the `auto` keyword making it pretty hard to parse, see https://github.com/bytedeco/javacpp/issues/407, but something like this should work, for now:
```java
conv = new Conv2dImpl(numChannelsIn, numChannelsOut, new LongPointer(1).put(3));
conv.stride().put(3);
conv.padding().put(3);
conv.bias().put(false);
```
Is this the only thing that's not working for you though? Instead of small usability issues like that, I would like to hear in priority about important issues preventing someone from getting anything working at all, and fix those first. However, if everything else is working for you, please let me know, and we can start looking at what we can to start polishing that.

**Comment by HGuillemet** [comment]
Ok, I didn't realize we could set options  this way. 
Continuing my porting attempt... 
There seems to be missing a direct way to create `TensorArrayRef` from a Java array of tensors, but I guess this could work:
```
new TensorArrayRef(new PointerPointer<Tensor>(x.length).put(x));
```

**Comment by HGuillemet** [comment]
There seems to be an issue with `Sequential` and `AnyModule`.

It's common in Pytorch to define a module as a sequence of other modules:
```
model = nn.Sequential(
          nn.Conv2d(1,20,5),
          nn.ReLU()
        )
```
which maps with the C++ API to:
```
torch::nn::Sequential model(
  torch::nn::Conv2d(1,20,5),
  torch::nn::ReLU()
);
```
A `Sequential` is basically a list of `AnyModule`, `AnyModule` being an abstract type representing any module. But because the `forward` method of modules may have any signature, usual inheritance is not an option. Calling the `forward` method of an `AnyModule` delegates the call to the underlying concrete module with dynamic checking of the signature.

In the preset we cannot rely on the C++ library and its implementation of AnyModule to dynamically call a `forward` method defined in Java (can we ?). 
I believe there are other cases where the C++ lib must call the `forward` method of a module. 
How to handle this ?

**Comment by saudet** [comment]
> Ok, I didn't realize we could set options this way.
> Continuing my porting attempt...
> There seems to be missing a direct way to create `TensorArrayRef` from a Java array of tensors, but I guess this could work:
> 
> ```
> new TensorArrayRef(new PointerPointer<Tensor>(x.length).put(x));
> ```

It looks like that needs an actual array of tensors, and not an array of pointers to tensors, so if we don't have that, we need to first create one possibly like this:
```java
Tensor array = new Tensor(x.length);
for (int i = 0; i < x.length; i++) {
    array.getPointer(i).put(x[i]);
}
new TensorArrayRef(array, x.length);
```

> A `Sequential` is basically a list of `AnyModule`, `AnyModule` being an abstract type representing any module. But because the `forward` method of modules may have any signature, usual inheritance is not an option. Calling the `forward` method of an `AnyModule` delegates the call to the underlying concrete module with dynamic checking of the signature.
> 
> In the preset we cannot rely on the C++ library and its implementation of AnyModule to dynamically call a `forward` method defined in Java (can we ?).
> I believe there are other cases where the C++ lib must call the `forward` method of a module.
> How to handle this ?

Yeah, that looks really complicated. I don't see at the moment how we could make this work nicely. It doesn't sound like the `Sequential` module is required for anything though. According to the docs, it's only there for convenience:
> ```
> /// Why should you use `Sequential` instead of a simple `std::vector`? The value
> /// a `Sequential` provides over manually calling a sequence of modules is that
> /// it allows treating the whole container *as a single module*, such that
> /// performing a transformation on the `Sequential` applies to each of the
> /// modules it stores (which are each a registered submodule of the
> /// `Sequential`). For example, calling
> /// `.to(torch::kCUDA)` on a `Sequential` will move each module in the list to
> /// CUDA memory. For example:
> ```

**Comment by HGuillemet** [comment]
New issue: the C++ lib handles modules either as normal struct/object (the `Module` class) or as shared references wrapped in instances of a generic class: `ModuleHolder<ModuleType>`. See https://pytorch.org/tutorials/advanced/cpp_frontend.html#module-ownership.
I don't think the module holder can work with the preset, at least for user Java modules, for the same reason that `Sequential`. So you might as well strip all related classes from the parsing, if possible.

The documentation says the use of module holders is needed for some features like the serizalization API. I don't know why and haven't tested it (I haven't found the entrypoint in the preset).

Another issue : the `register_module` method has been declined for all concrete class of Module in the lib. But it should accept any shared_ptr to a Module subclass, so a single definition `register_module(String, Module)` should work and will allow to register a custom module. More generally,  `std::shared_ptr<ModuleType>` could be parsed into `Module` everywhere.

**Comment by saudet** [comment]
> New issue: the C++ lib handles modules either as normal struct/object (the `Module` class) or as shared references wrapped in instances of a generic class: `ModuleHolder<ModuleType>`. See https://pytorch.org/tutorials/advanced/cpp_frontend.html#module-ownership.
> I don't think the module holder can work with the preset, at least for user Java modules, for the same reason that `Sequential`. So you might as well strip all related classes from the parsing, if possible.
> 
> The documentation says the use of module holders is needed for some features like the serizalization API. I don't know why and haven't tested it (I haven't found the entrypoint in the preset).

Actually, we *can* use `ModuleHolder`, it just doesn't lead to nice code in Java, but it does work, so I wouldn't worry about that for now, at least unless it actually doesn't work for some functions and prevents you from doing something.

> Another issue : the `register_module` method has been declined for all concrete class of Module in the lib. But it should accept any shared_ptr to a Module subclass, so a single definition `register_module(String, Module)` should work and will allow to register a custom module. More generally, `std::shared_ptr<ModuleType>` could be parsed into `Module` everywhere.

Ah, yes, good catch. I've added that to commit https://github.com/bytedeco/javacpp-presets/commit/2d940f004516e42565e025562a6955b1aa9008f7. Thanks!

**Comment by HGuillemet** [comment]
> Actually, we _can_ use `ModuleHolder`, it just doesn't lead to nice code in Java, but it does work,

Can you create a `ModuleHolder` for your java custom module?

**Comment by saudet** [comment]
I should be possible now with the latest changes, yes.

**Comment by HGuillemet** [comment]
A detail : `ConvXdTransposeOptions` and `ConvXdTransposeFuncOptions` should be called `ConvTransposeXdOptions` and `ConvTransposeXdFuncOptions` for consistency with other names and with the C++ lib own alias defined in `conv.h`:
```c++
using ConvTranspose2dOptions = ConvTransposeOptions<2>;
```

**Comment by HGuillemet** [comment]
> Ah, yes, good catch. I've added that to commit [2d940f0](https://github.com/bytedeco/javacpp-presets/commit/2d940f004516e42565e025562a6955b1aa9008f7). Thanks!

Thanks, I see the new `register_module(String, Module)` but the specialized versions are still there. Is this wanted ?

**Comment by saudet** [comment]
> A detail : `ConvXdTransposeOptions` and `ConvXdTransposeFuncOptions` should be called `ConvTransposeXdOptions` and `ConvTransposeXdFuncOptions` for consistency with other names and with the C++ lib own alias defined in `conv.h`:
> 
> ```c++
> using ConvTranspose2dOptions = ConvTransposeOptions<2>;
> ```

Oops, fixed in commit https://github.com/bytedeco/javacpp-presets/commit/cfb9c6adb32e7a484f09b667a93008d2def3f878. Thanks!

> > Ah, yes, good catch. I've added that to commit [2d940f0](https://github.com/bytedeco/javacpp-presets/commit/2d940f004516e42565e025562a6955b1aa9008f7). Thanks!
> 
> Thanks, I see the new `register_module(String, Module)` but the specialized versions are still there. Is this wanted ?
> If this is for keeping the return type, what about:
> 
> ```
> public native <T extends Module> T register_module(String name, T module);
> ```

We can add a helper method like that, sure, but the template itself does a `std::dynamic_pointer_cast<ModuleType>()` that may return with a different address from its argument. Since it's not difficult to generate them, we might as well leave them there.

**Comment by HGuillemet** [comment]
```
Conv2dImpl conv = new Conv2d(...);
register_module("conv", conv);
System.err.println(modules(false).get(0).getClass());
```
We get `org.bytedeco.pytorch.Module`. That's a bit annoying. Do you think we can do something about it ?
How are the shared pointer handled ? Is there a way to get the original Java object when the C++ lib returns a shared pointer ?
Something like a weak hash map Pointer => Object managed by JavaCPP, or something more specific to Pytorch preset that would adapt the ModuleHolder mechanism ?

**Comment by HGuillemet** [comment]
I'd like to use tensor indexing like `[:,0]`. I need to allocate a `TensorIndex` of size 2, fill it with `new TensorIndex(new Slice())` and `new TensorIndex(0)` and wrap a TensorIndexArrayRef around it. Right ?
However I don't find a way to create a `TensorIndex` of size 2 since `Pointer(long)` is overridden by the single-index  constructor.

**Comment by HGuillemet** [comment]
Any way to use `Module.pretty_print` that takes a `std::ostream` ?

**Comment by saudet** [comment]
> ```
> Conv2dImpl conv = new Conv2d(...);
> register_module("conv", conv);
> System.err.println(modules(false).get(0).getClass());
> ```
> 
> We get `org.bytedeco.pytorch.Module`. That's a bit annoying. Do you think we can do something about it ?
> How are the shared pointers handled ? Is there a way to get the original Java object when the C++ lib returns a shared pointer ?
> Something like a map Pointer => Object managed by JavaCPP, or something more specific to Pytorch preset that would adapt the ModuleHolder mechanism ?

It's not possible to hold references to arbitrary Java objects in C++ data structures. Of course you could start hacking with JNI directly, but then you wouldn't be coding in Java anyways. To do that kind of thing in Java, we need to use something like a `HashMap<Module, YourModuleSubclass>`. JavaCPP could offer some functionality to do that more implicitly, see issue https://github.com/bytedeco/javacpp/issues/419 for some discussions, but we can do it easily enough manually, so I don't consider it a priority.

JavaCPP keeps pointers to containers like `std::shared_ptr<>` in the `Pointer.NativeDeallocator.ownerAddress` field, which is what it uses to deallocate.

> I'd like to use tensor indexing like `[:,0]`. I need to allocate a `TensorIndex` of size 2, fill it with `new TensorIndex(new Slice())` and `new TensorIndex(0)` and wrap a TensorIndexArrayRef around it. Right ?
> However I don't find a way to create a `TensorIndex` of size 2 since `Pointer(long)` is overridden by the single-index constructor.

I think it's more like `new TensorIndexArrayRef(new TensorIndex(new Slice(null, new LongOptional(0), null)))`, isn't?

> Any way to use `Module.pretty_print` that takes a `std::ostream` ?

Hum, there's no overload that works with strings, this is annoying. I've added a way to fetch pointers to `std::cout`, `std::cerr`, and `std::clog` in commit https://github.com/bytedeco/javacpp-presets/commit/acfb6a2570030cd7b61dc9d566f865a09dd77979. I hope that's enough for your needs?

**Comment by HGuillemet** [comment]
> 
> > I'd like to use tensor indexing like `[:,0]`. I need to allocate a `TensorIndex` of size 2, fill it with `new TensorIndex(new Slice())` and `new TensorIndex(0)` and wrap a TensorIndexArrayRef around it. Right ?
> > However I don't find a way to create a `TensorIndex` of size 2 since `Pointer(long)` is overridden by the single-index constructor.
> 
> I think it's more like `new TensorIndexArrayRef(new TensorIndex(new Slice(null, new LongOptional(0), null)))`, isn't?

I don't think so. I meant to return every first colum of all rows. I believe I need 2 `TensorIndex`.

> > Any way to use `Module.pretty_print` that takes a `std::ostream` ?
> 
> Hum, there's no overload that works with strings, this is annoying. I've added a way to fetch pointers to `std::cout`, `std::cerr`, and `std::clog` in commit [acfb6a2](https://github.com/bytedeco/javacpp-presets/commit/acfb6a2570030cd7b61dc9d566f865a09dd77979). I hope that's enough for your needs?

The `pretty_print` doesn't print anything useful and doesn't do the recursion for some reason probably related to the module reference issue. I'll investigate. But the new `cerr()` works, thanks.
Would it make sense to have a preset for the iostream library ?

**Comment by saudet** [comment]
> I don't think so. I meant to return every first colum of all rows. I believe I need 2 `TensorIndex`.

I see, it looks like we're supposed to use `std::vector<at::indexing::TensorIndex>` for this so I've added that in commit https://github.com/bytedeco/javacpp-presets/commit/9210a49aa8427d7c2dd56ec3ff98843ed2cc9f3b.

> The `pretty_print` doesn't print anything useful and doesn't do the recursion for some reason probably related to the module reference issue. I'll investigate. But the new `cerr()` works, thanks.

It looks like we need to call `operator<<()` for this, but its declaration is missing so I've added that in commit https://github.com/bytedeco/javacpp-presets/commit/9210a49aa8427d7c2dd56ec3ff98843ed2cc9f3b as well.

> Would it make sense to have a preset for the iostream library ?

Possibly, but since it's not a really great API, most libraries offer overloads using strings instead anyway.

**Comment by HGuillemet** [comment]
When I'm trying to use `Tensor.data_ptr_long()`, with CUDA enabled I get the following linking error:
```
linux-x86_64-gpu/libjnitorch.so: undefined symbol: _ZNK2at6Tensor8data_ptrIxEEPT_v
```
`data_ptr_int` or `date_ptr_float` does work. 
Maybe it's not directly related to the preset, but `data_ptr` in a quick C++ program seems to work with a long tensor.
Any idea what might cause this or should I report upstream ?

**Comment by HGuillemet** [comment]
`cuda.h` may be added to the parsed files, to support for instance `torch::cuda::is_available`. 
However if we'd better avoid the name `is_available` as is in the global Java `torch` class. 
Would it be worth to try to reconstruct the namespaces hierarchy in Java for this preset ?

**Comment by saudet** [comment]
It appears that we need to use `int64_t` as 64-bit scalar type. I've fixed that in commit https://github.com/bytedeco/javacpp-presets/commit/622bd5b3d4dcbd3b5d2d853ac3682861ec3a04a1. Thanks!

> `cuda.h` may be added to the parsed files, to support for instance `torch::cuda::is_available`.
> However if we'd better avoid the name `is_available` as is in the global Java `torch` class.
> Would it be worth to try to reconstruct the namespaces hierarchy in Java for this preset ?

I've added that in commit https://github.com/bytedeco/javacpp-presets/commit/622bd5b3d4dcbd3b5d2d853ac3682861ec3a04a1 as well and simply prefixed those functions with `cuda_`. Looks good?

**Comment by HGuillemet** [comment]
It does work now, thanks.

Another point : in the C++ API, the Python equivalent to `with no_grad():` is the `NoGradGuard` thread local. 
Can you think of a way to have this feature from Java ?

**Comment by HGuillemet** [comment]
I now have something trainable. But not on GPU yet.
This code:
```Java
      Tensor input = torch.empty(10, 1, 256, 256);
      input.to(new Device(kCUDA()));
```
Throws:
```
java.lang.RuntimeException: Unsupported TypeMeta in ATen:  (please report this error)
Exception raised from error_unsupported_typemeta at ../c10/util/typeid.cpp:22 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x68 (0x7f105405be78 in /home/hg/.javacpp/cache/pytorch-1.8.1-1.5.6-SNAPSHOT-linux-x86_64-gpu.jar/org/bytedeco/pytorch/linux-x86_64-gpu/libc10.so)
```
(note the empty Data Type).

I have also got SIGSEGV with  similar code.
The equivalent in C++:
```C++
torch::Tensor foo = torch::empty({10, 1, 256, 256});
torch::Tensor foo2 = foo.to(c10::Device(c10::kCUDA));
```
Seems to work.
Any idea ?

**Comment by saudet** [comment]
> Another point : in the C++ API, the equivalent of Python `with no_grad()` is the `NoGradGuard` thread local.
> Can you think of a way to have this feature from Java ?

That should work pretty much as is, but please let me know if something like this doesn't work for some reason:
```java
try (NoGradGuard no_grad = new NoGradGuard()) {
    // ...
}
```


> The equivalent in C++:
> 
> ```c++
> torch::Tensor foo = torch::empty({10, 1, 256, 256});
> torch::Tensor foo2 = foo.to(c10::Device(c10::kCUDA));
> ```
> 
> Seems to work.
> Any idea ?

It looks like that's implicit creating TensorOptions, so we would need to do the same in Java. This seems to work fine:
```java
Tensor foo = empty(10, 1, 256, 256);
Tensor foo2 = input.to(new TensorOptions().device(new DeviceOptional(new Device(kCUDA()))), false, false, null);
```

**Comment by HGuillemet** [comment]
> That should work pretty much as is, but please let me know if something like this doesn't work for some reason:
> 
> ```java
> try (NoGradGuard no_grad = new NoGradGuard()) {
>     // ...
> }
> ```

Oops sorry, it does work.

> It looks like that's implicitly creating TensorOptions, so we would need to do the same in Java. This seems to work fine:
> 
> ```java
> Tensor foo = empty(10, 1, 256, 256);
> Tensor foo2 = foo.to(new TensorOptions().device(new DeviceOptional(new Device(kCUDA()))), false, false, null);
> ```

Ok. This works too:
```
foo2 = foo.to(new Device(kCUDA()), ScalarType.Float);
```
But this does not:
```
foo2 = foo.to(new TensorOptions().device(new DeviceOptional(new Device(kCUDA()))));
```
So it seems the problem is related to the specification of the data type.

**Comment by saudet** [comment]
It's just calling the wrong overload since `caffe2::TypeMeta` is mapped to `Pointer`. I've adjusted that more properly in commit https://github.com/bytedeco/javacpp-presets/commit/50c3dbb5bac6a6af6ec5bf8ba0f9b17fa9fa19ec so you won't be able to do that anymore. I've also created an instance for `c10::TensorOptions<c10::Device>()` so we can shorten the call to:
```java
Tensor foo2 = foo.to(new TensorOptions(new Device(kCUDA())), false, false, null);
```

**Comment by HGuillemet** [comment]
Here is a summary of my experience with the preset.
I have completed the porting of a somewhat complex CNN for detection and tracking object and it works well, quite fast. 
Compared to Python, Java makes its job of increasing rigor, readability  and safety of the code, while being more IDE friendly.
I'll most certainly use this preset for my DL projects in Java from now on.

What I have tested:
* Definition of a CNN model
* Training
* Inferring
* Save and load network weights and optimizer state
* Use of CPU and GPU

What I have not tested is the data loading part of the library. I don't think there is something really usable there since the API is heavily generic and many classes and functions were not parsed. For Tensor reading and writing, I used `data_ptr`, sometimes through an OpenCV `Mat`.

The single consequence of the fact that a module logic is implemented in Java (in a `forward` method but that could have any name) and that this method is not callable from the library is that `Sequential` block cannot be used (see [above](https://github.com/bytedeco/javacpp-presets/issues/623#issuecomment-814107543). But this is not a big deal. Sequential can be replaced by explicit, more Java-ish, custom modules.

I made no explicit use of the [module holder mechanism](https://pytorch.org/tutorials/advanced/cpp_frontend.html#module-ownership), and always call the `*Impl` version of modules. Is there any interest of using it from the preset ? If not, the related classes and methods could be hidden to rationalize the API.

Of course there is some constructs that are cumbersome to use as is and would benefit from some helper methods or classes in the preset or in a higher level library like mentioned above (TensorIndexArrayRef, Expanding Array...), but nothing blocking.

Also some more headers could probably be parse. The only one I can think about right now is `CUDAContext.h` to access the device properties and the available memories.

**Comment by saudet** [comment]
> Here is a summary of my experience with the preset.
> I have completed the porting of a somewhat complex CNN for detection and tracking object and it works well, quite fast.
> Compared to Python, Java makes its job of increasing rigor, readability and safety of the code, while being more IDE friendly.
> I'll most certainly use this preset for my DL projects in Java from now on.
> 
> What I have tested:
> 
> * Definition of a CNN model
> * Training
> * Inferring
> * Save and load network weights and optimizer state
> * Use of CPU and GPU

Thanks for the feedback and for your help getting this working! I'm happy to hear it's usable enough to build something with it.
 
> What I have not tested is the data loading part of the library. I don't think there is something really usable there since the API is heavily generic and many classes and functions were not parsed. For Tensor reading and writing, I used `data_ptr`, sometimes through an OpenCV `Mat`.

Actually, the `ChunkDataReader` and `ChunkDataset` classes seem to offer a pretty generic but dynamic interface using virtual functions, so it looks like something that we could map easily enough to Java. If these classes seem like something you'd like to try to use, let me know and I'll map them with virtual methods that we can implement in Java.

> The single consequence of the fact that a module logic is implemented in Java (in a `forward` method but that could have any name) and that this method is not callable from the library is that `Sequential` block cannot be used (see [above](https://github.com/bytedeco/javacpp-presets/issues/623#issuecomment-814107543). But this is not a big deal. Sequential can be replaced by explicit, more Java-ish, custom modules.
> 
> I made no explicit use of the [module holder mechanism](https://pytorch.org/tutorials/advanced/cpp_frontend.html#module-ownership), and always call the `*Impl` version of modules. Is there any interest of using it from the preset ? If not, the related classes and methods could be hidden to rationalize the API.

I'm not sure how we could make good use of all this, but I feel it needs further investigation before dismissing it entirely.

> No memory leak found, thanks to `PointerScope`.
> 
> Of course there is some constructs that are cumbersome to use as is and would benefit from some helper methods or classes in the preset or in a higher level library like mentioned above (TensorIndexArrayRef, Expanding Array...), but nothing blocking.
> 
> Also some more headers could probably be parsed. The only one I can think about right now is `CUDAContext.h` to access the device properties and the available memory.

The functions in  `CUDAContext.h` are not available in non-CUDA builds, so we would need to skip generating JNI for those functions in those builds, which we could do, but there's probably some other higher-level functions available that can return the information you're after. If you're just after `cudaDeviceProp`, we can get that from the CUDA API itself with this function:
http://bytedeco.org/javacpp-presets/cuda/apidocs/org/bytedeco/cuda/global/cudart.html#cudaGetDeviceProperties-org.bytedeco.cuda.cudart.cudaDeviceProp-int-

In any case, if you find any other classes or functions that you would like to have access to, please let me know!

**Comment by jbaron** [comment]
Did also a bit of experimentation in combination with Kotlin , some of my findings and questions:

1. Easy to use and no unexpected results. For example tried to bind some Tensor operations to Kotlin extension functions and it was straight forward. Just see https://gist.github.com/jbaron/8b770759638b4ee79eaf7ec0f09364ea 

2. CNN and fully connected network architectures work all as expected.

3. Not sure how to destructure LSTM result into its 3 values. But I guess this is more an user (me) issue than an issue with javacpp ;)

4. Still a bit confused if I when I have to to call close()   For example in the Tensor.get in the above gist, should I call close on all allocations?

5. Some of the method names conflict with the Kotlin operator names, making it impossible to implement extension functions with operator overloading. Of course when wrapping Tensor into a custom class, this goes away.  

6. If I understood correctly, the trace functionality is implemented in Python and not C++. Came as a bit of a surprise to me (but again nothing to do wit the javacpp binding).

**Comment by saudet** [comment]
Thanks for the feedback @jbaron! I'm happy to hear that it's usable enough for you as well.

> Did also a bit of experimentation in combination with Kotlin , some of my findings and questions:
> 
> 1. Easy to use and no unexpected results. For example tried to bind some Tensor operations to Kotlin extension functions and it was straight forward. Just see https://gist.github.com/jbaron/8b770759638b4ee79eaf7ec0f09364ea
> 
> 2. CNN and fully connected network architectures work all as expected.
> 
> 3. Not sure how to destructure LSTM result into its 3 values. But I guess this is more an user (me) issue than an issue with javacpp ;)

Which result is that exactly? There might be issues with the mapping obviously.

> 4. Still a bit confused if I when I have to to call close()   For example in the Tensor.get in the above gist, should I call close on all allocations?

Whenever in doubt, you can call `close()` on everything, it won't do anything bad. For simplicity though, we can use `PointerScope` with a try-with-resources statement as shown here: http://bytedeco.org/news/2018/07/17/bytedeco-as-distribution/

> 5. Some of the method names conflict with the Kotlin operator names, making it impossible to implement extension functions with operator overloading. Of course when wrapping Tensor into a custom class, this goes away.
> 
> 6. If I understood correctly, the trace functionality is implemented in Python and not C++. Came as a bit of a surprise to me (but again nothing to do wit the javacpp binding).

It sounds like it would be a huge endeavor indeed: https://github.com/pytorch/pytorch/issues/26086#issuecomment-532321136

**Comment by jbaron** [comment]
Regarding point 3, I cannot get the returned result of a LSTM forward, back to its 3 core elements:


```

// input is shape (20, 16, 10) and by not providing extra params to the forward call, 
// I'm using implicit the zero init for the hidden state and cell state weights.
val x = lstm1.forward(input)

 // return is of type  tuple(output, tuple(hidden_state, cell_state))

```
 
So how to get x into its three elements: output, hidden_state and cell_state. Tried a few approaches with the PyTorch Tuple class, but they didn't work out.

**Comment by saudet** [comment]
@jbaron I see, those were simply not mapped. I've fixed that in commit https://github.com/bytedeco/javacpp-presets/commit/41f5aee18f2f95be2b14a4e8609fda76c9cebf61.
Let me know if you're having problems with those! Thanks

**Comment by jbaron** [comment]
Did a quick test and LSTM now seem to be working fine, thanks!!!

What I used as as a simple test model (is Kotlin code):

```kotlin
class RecurrentNet : Module() {

    private val lstm1 = register_module("lstm1", LSTMImpl(50, 20))
    private val fc1 = register_module("fc1", LinearImpl(20, 10))

    fun forward(t: Tensor): Tensor {
        val x = lstm1.forward(t)
        
        // x[:, -1, :]
        val y = x.get0().index(TensorIndexArrayRef(TensorIndexVector(
            TensorIndex(Slice()),
            TensorIndex(-1),
            TensorIndex(Slice())
        )))
        return torch.log_softmax(fc1.forward(y), 1)
    }

}

```

**Comment by jbaron** [comment]
Having some troubles using the base class Module in some scenario's. 

For example the following line of code crashes the JVM:

```

val lstm = LSTMImpl(50, 20)
register_module("lstm1", lstm as Module)

```

This is not really the code I need, but just serves as a simple example to replicate the unexpected behavior. Any ideas what is going on?

**Comment by saudet** [comment]
If a reference to `lstm` isn't kept in a field, it may be getting GCed.

**Comment by jbaron** [comment]
Perhaps not clear from my description, but it is the "as Module" cast that makes it crash. Without it, it is working fine. 
So that should not be related to GC, right?

**Comment by saudet** [comment]
`register_module()` is a function template, and looking at its definition, it's going to need to know the type of your `Module` at compile time to be able to "move" it correctly: https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/module.h#L650

**Comment by jbaron** [comment]
Thanks, will have to look into this in a bit more detail I guess (especially also for own defined Modules and the implications on how to use/register them). Time to brush up on my C/C++ skills ;)

**Comment by HGuillemet** [comment]
In my case I haven't face any error when calling `register_module` passing an argument with a type being 
* the concrete class of a predefined library module, 
* the concrete class of a Java custom module extending `Module`
* `Module`, when the module is a custom module.

**Comment by jbaron** [comment]
Indeed seems that if I play by those rules no issue, thanks. 

P.S from my understanding the main reason to register a module is to make it easier to retrieve the trainable parameters. I guess some dictionary traversal is performed when you call net.parameters(). But is you could get the parameters in another way, it is not really required. But I could be very wrong here.

**Comment by saudet** [comment]
Actually, I think the problem is with the dynamic cast here below:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/module.h#L651
It's not usually a problem what the input type is, but I didn't realize there was virtual inheritance involved, then it matters:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/cloneable.h#L24
We can probably make this work by providing calls to `dynamic_cast<Module>(LSTMImpl*),` etc, which would return correct pointers to the base class `Module`. @jbaron Is this something you need?

**Comment by jbaron** [comment]
@saudet  thanks for looking in to this. Right now only doing some simple prototyping to see what could be the best approach for my use-cases. So nothing is really vital at this moment, especially if there is a work around like in this case.

**Comment by saudet** [comment]
@jbaron I've updated to PyTorch 1.9.0 in commit https://github.com/bytedeco/javacpp-presets/commit/f1ea76f625fb74476d90912be0eb451037b651b7, and I also added `Module.asModule()` methods to upcast correctly and do things like you wanted like this:
```java
var lstm = new LSTMImpl(50, 20);
register_module("lstm1", lstm.asModule());
```
Let me know if there's anything else missing that you would like to have! And thanks for testing

**Comment by HGuillemet** [comment]
With the update, the data type for padding has changed from an expanding array of longs to some variant type that can be either an array, kSame or kValid. Code like:
```
convOpt.padding.put(new long[] { 3, 3 });
```
doesn't work anymore.

How can we handle this variant type ?

**Comment by jbaron** [comment]
Would something like this not work (Koltin sample):

 ```
val lp =  LongPointer()
 lp.put(3L, 3L)
 convOpt.padding().put<LongPointer>(lp)
```

**Comment by HGuillemet** [comment]
There must be some header value discriminating the actual type of the value stored in the variant.

Also beware of `put(long, long)`: it's a a call to the method that puts a single value at a specific index. It will most probably give you a segmentation fault.

**Comment by jbaron** [comment]
Thanks, indeed misread the LongPointer API.  The following code runs without a segmentation fault:

```

import org.bytedeco.javacpp.LongPointer
import org.bytedeco.pytorch.*
import org.bytedeco.pytorch.global.torch

class Net : Module() {

    private val conv = register_module("conv", getConv())

    private fun getConv(): Conv2dImpl {
        val kernel = LongPointer(2)
        kernel.put(0L, 3L)
        kernel.put(1L, 3L)

        val convOpt = Conv2dOptions(3L, 16L, kernel)
        val padding = LongPointer(2)
        padding.put(0L, 3L)
        padding.put(1L, 3L)
        convOpt.padding().put<LongPointer>(padding)
        return  Conv2dImpl(convOpt)

    }

    fun forward(t: Tensor): Tensor {
        return conv.forward(t)
    }

}

fun main() {
    val net = Net()
    val x = torch.rand(16,3, 28, 28)
    net.forward(x)
}


```

**Comment by saudet** [comment]
Right, we probably need to map those variants to set the options correctly, so I've done that in commit https://github.com/bytedeco/javacpp-presets/commit/7b9ccf3c70ccb76508824b785773ed008ae084f1. Please give it a try!

**Comment by HGuillemet** [comment]
It now works as with version 1.8 for my cases with CNN and explicit paddings. Thanks !

**Comment by jbaron** [comment]
Short question, when a method expects a pointer to std::ostream, how to provide this? Are these C++ std lib classes also mapped to Java equivalents?

**Comment by saudet** [comment]
> Short question, when a method expects a pointer to std::ostream, how to provide this? Are these C++ std lib classes also mapped to Java equivalents?

It's not currently mapped to anything, it's a pretty ugly API, so usually we get overloads using `std::string`, but PyTorch doesn't do that for some reason. It would require some serious considerations to get something usable in Java, but anyway we can get pointers to `std::cout`, `std::cerr`, and `std::clog` if that's all you need, see above https://github.com/bytedeco/javacpp-presets/issues/623#issuecomment-817299627.

**Comment by saudet** [comment]
The presets for the C++ API of PyTorch 1.9.0 have been released with version 1.5.6! Thanks to everyone for helping out with this effort. I'll close this issue, but please do open new separate threads to continue the discussion about various remaining issues and potential enhancements.

/cc @stu1130

**Comment by saudet** [comment]
AFAIK, `torch::OrderedDict<std::string,at::Tensor>` is pretty much the same as `std::map<std::string,at::Tensor>`, so the key is a string, that sounds alright, and it's possible that it can't be safely iterated, yes. Do you have counterexamples, in C++?

**Comment by HGuillemet** [comment]
Back to this issue.
I realized that `ordering_dict.h` is not parsed and `OrderedDict` is declared as a `basic/container` instead.
Thus `OrderingDict` specific features, for instance an overload of `[]`, taking an integer, is not available.
I guess there is no mean to have `ordereding_dict.h` parsed while keeping the generic Dictionary features like `get(key)` and iterators ? Is it ?

About iterators: they seem to work as expected using this pattern :
```
for (Iterator it=dict.begin(); !it.equals(dict.end()); it = it.increment()) {
  ...
}
```

**Comment by saudet** [comment]
It should be possible to parse `ordered_dict.h`, it doesn't look like the kind of really complicated template that you'd find in STL that are essentially unparsable by anything else than a full C++ compiler. But is there anything that you're not able to do with the current mapping?

**Comment by HGuillemet** [comment]
Not at the moment, nothing prioritary.

**Comment by saudet** [comment]
@HGuillemet I've included `ordered_dict.h` in commit https://github.com/bytedeco/javacpp-presets/commit/78540be41638b6b19bcc12982152d5beba464984 to map `torch::OrderedDict<std::string,at::Tensor>` like you wanted. Please give it a try with the snapshots: http://bytedeco.org/builds/

Let  me know if there is anything else missing!
